### PR TITLE
Fix graceful shutdown with in flight client requests

### LINKED ISSUE #2539: Router doesn't seem to gracefully handle in flight requests on shutdown
**Describe the bug**
Router doesn't seem to gracefully handle in flight requests on shutdown.

**To Reproduce**
1. Start router (in dev or release mode)
2. Hit a resolver that has a 15s sleep in it
3. Stop the router
4. See the in flight request fail

**Expected behavior**
Request should go through, and the router should wait for a given period (would be nice if that matched the supergraph timeout configuration) for all requests to drain.

### PR DESCRIPTION
**Pull request by Geal**
Fix #2539
Blocked by #2666

### CODE REVIEW

File: apollo-router/src/state_machine.rs [hunk 1]
```diff
@@ -183,13 +185,19 @@ impl<FA: RouterSuperServiceFactory> State<FA> {
...
             } => {
                 tracing::info!("shutting down");
-                server_handle
+                let state = server_handle
                     .shutdown()
                     .map_ok_or_else(Errored, |_| Stopped)
-                    .await
+                    .await;
+                tracing::info!("server was shut down");
+                //FIXME: we might want to set a timeout here
```
Comment (bnjjj) [inline apollo-router/src/state_machine.rs:197]: Should it be part of this PR ?

Comment (Geal) [reply apollo-router/src/state_machine.rs:197]: no, I'll remove it

**Comment by Geal** [comment]
good news, it does not seem to affect the perf tests

File: apollo-router/tests/lifecycle_tests.rs [hunk 2]
```diff
@@ -79,3 +85,61 @@ async fn create_router(config: &str) -> Result<IntegrationTest, BoxError> {
...
+        let (_, res) = router.run_query().await;
+        (
+            router,
+            serde_json::from_slice::<graphql::Response>(&res.bytes().await.unwrap()).unwrap(),
+        )
+    });
+
+    tokio::time::sleep(Duration::from_millis(1000)).await;
+    unsafe {
+        libc::kill(pid, libc::SIGINT);
```
Comment (Meemaw) [inline apollo-router/tests/lifecycle_tests.rs:134]: Just making sure -- it will also work with SIGTERM, right?

**Review by bnjjj — APPROVED** [review]

File: apollo-router/tests/lifecycle_tests.rs [hunk 3]
```diff
@@ -103,130 +88,56 @@ async fn create_router(config: &str) -> Result<IntegrationTest, BoxError> {
...
+    }
+    tokio::time::sleep(Duration::from_millis(1000)).await;
 
-    let data = client_handle.await.unwrap();
+    let (mut router, data) = client_handle.await.unwrap();
     insta::assert_json_snapshot!(data);
+    router.assert_shutdown().await;
 
     println!("executed client request");
```
Comment (BrynCooke) [inline apollo-router/tests/lifecycle_tests.rs:142]: NIT, println

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/tests/lifecycle_tests.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/tests/lifecycle_tests.rs:134]: the router considers both SIGINT and SIGTERM: https://github.com/apollographql/router/blob/dev/apollo-router/src/router.rs#L614-L623

**Pull request merged**